src/Apps/Apps.jl

```diff
@@ -0,0 +1,403 @@
+            continue
+        end
+        Pkg.activate(joinpath(APP_ENV_FOLDER, info.name)) do
+            @info "Precompiling $(info.name)..."
+            Pkg.precompile()
+        end
+    end
+end
+
+function require_not_empty(pkgs, f::Symbol)
+    pkgs === nothing && return
+    isempty(pkgs) && pkgerror("app $f requires at least one package")
+end
+
+function rm(pkgs_or_apps::Union{Vector, Nothing})
+    if pkgs_or_apps === nothing
+        rm(nothing)
+    else
+        for pkg_or_app in pkgs_or_apps
+            if pkg_or_app isa String
+                pkg_or_app = PackageSpec(pkg_or_app)
+            end
+            rm(pkg_or_app)
+        end
+    end
+end
+
+function rm(pkg_or_app::Union{PackageSpec, Nothing}=nothing)
+    pkg_or_app = pkg_or_app === nothing ? nothing : pkg_or_app.name
+
+    require_not_empty(pkg_or_app, :rm)
+
+    manifest = Pkg.Types.read_manifest(joinpath(APP_ENV_FOLDER, "AppManifest.toml"))
+    dep_idx = findfirst(dep -> dep.name == pkg_or_app, manifest.deps)
+    if dep_idx !== nothing
+        dep = manifest.deps[dep_idx]
+        @info "Deleted all apps for package $(dep.name)"
+        delete!(manifest.deps, dep.uuid)
+        for (appname, appinfo) in dep.apps
+            @info "Deleted $(appname)"
+            Base.rm(joinpath(JULIA_BIN_PATH, appname); force=true)
+        end
+        Base.rm(joinpath(APP_ENV_FOLDER, dep.name); recursive=true)
+    else
+        for (uuid, pkg) in manifest.deps
+            app_idx = findfirst(app -> app.name == pkg_or_app, pkg.apps)
+            if app_idx !== nothing
+                app = pkg.apps[app_idx]
+                @info "Deleted app $(app.name)"
+                delete!(pkg.apps, app.name)
+                Base.rm(joinpath(JULIA_BIN_PATH, app.name); force=true)
+            end
+            if isempty(pkg.apps)
+                delete!(manifest.deps, uuid)
+                Base.rm(joinpath(APP_ENV_FOLDER, pkg.name); recursive=true)
+            end
+        end
+    end
+
+    Pkg.Types.write_manifest(manifest, APP_MANIFEST_FILE)
+    return
+end
+
+
+
+#########
+# Shims #
+#########
+
+function generate_shims_for_apps(pkgname, apps, env)
+    for (_, app) in apps
+        generate_shim(app, pkgname; env)
+    end
+end
+
+function generate_shim(app::AppInfo, pkgname; julia_executable_path::String=joinpath(Sys.BINDIR, "julia"), env=joinpath(homedir(), ".julia", "environments", "apps", pkgname))
+    filename = joinpath(homedir(), ".julia", "bin", app.name * (Sys.iswindows() ? ".bat" : ""))
+    mkpath(dirname(filename))
+    content = if Sys.iswindows()
+        windows_shim(pkgname, julia_executable_path, env)
+    else
+        bash_shim(pkgname, julia_executable_path, env)
+    end
+    overwrite_if_different(filename, content)
+    if Sys.isunix()
+        chmod(filename, 0o755)
+    end
+end
+
+
+function bash_shim(pkgname, julia_executable_path::String, env)
+    return """
+        #!/usr/bin/env bash
```

Comment:
On a briefer note, shouldn't the unix shim be based on `sh` not `bash`, possibly even `#!/bin/sh` over `#!/usr/bin/env sh` (somebody else should check, but IIRC that's the POSIX form)?

Reply:
"Somebody else" is apparently me :smile:, and my "IIRC" was wrong.

I've just had a look at https://pubs.opengroup.org/onlinepubs/009695399/utilities/sh.html and under **APPLICATION USAGE** there's this relevant excerpt:

> Applications should note that the standard PATH to the shell cannot be assumed to be either /bin/sh or /usr/bin/sh, and should be determined by interrogation of the PATH returned by [getconf](https://pubs.opengroup.org/onlinepubs/009695399/utilities/getconf.html) PATH , ensuring that the returned pathname is an absolute pathname and not a shell built-in.

Reply:
From reading around a bit more, it seems like the consensus on portability goes something like this (from most portable to least):
- A binary executable that doesn't have a shebang
- `#!/bin/sh`
- `#!/usr/bin/env sh`
- `#!/usr/bin/env bash`
- `#!/bin/bash`